### PR TITLE
feat(perf): Document span operations spec

### PR DESCRIPTION
**Pull request by AbhiPrasad**
Edit: Please note, the scope of this PR has changed since it's creation. After reading the PR description, please see: https://github.com/getsentry/develop/pull/452#issuecomment-983716185

This PR adds documentation around the span operations spec that we use.

# Background

Before reviewing, I would recommend reading through the `Introduction` of this document: [https://develop.sentry.dev/sdk/research/performance#introduction](https://develop.sentry.dev/sdk/research/performance#introduction)

`OpenTracing`, which we based our initial performance product off of, had a concept of operation name. We can see it being leveraged in this [early JS SDK PR](https://github.com/getsentry/sentry-javascript/pull/1918/files#diff-b7d385e8e3c69cd4ade2edd41436b21255dde1d851717dc5c22cc7623a911a37R61).

According to OpenTracing (which became OpenTelemetry btw): [https://opentracing.io/specification/](https://opentracing.io/specification/)

As we moved forward, we introduced [both the operation and description fields to spans](https://github.com/getsentry/sentry-python/pull/372/files#diff-771cb012ea65fcb826b77e2abb36b504fdf3407fbe8054857b05fe5a41ed5164R80). Now, these seem redundant (and in fact OpenTelemetry just has [span name](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span)) - but if we use them correctly, they can be an advantage in our spec.

image

From the current usage of these attributes, span operation seems to typically be a low cardinality value, a short code that provides instant understanding to the user when they see a span. For example, `db` or `http`. Descriptions tend to be way more high cardinality, so a `db` operation would have the SQL query as it's description, while a `http` operation would have the resource URL as it's description.

Although these conventions were never codified or formally specced out, they seem to be providing a high degree of customer value. As such, when the span operations spec was created, it was heavily based on the existing conventions.

# What was done

Please read through `src/docs/sdk/performance/span-operations.mdx`.

In short, we introduce the ideas of categories and actions when defining span operations. Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers.

# Next Steps

For now, this PR heavily focuses on defining the span operations. As a next step, we may start taking a look at [span descriptions](#463) or other metadata. Eventually, we want to establish a set of semantic conventions around performance monitoring, similar to OpenTelemetry's. In this case, we probably want to rely on OpenTelemetry's semantic conventions for backend focused use cases. For areas like desktop, mobile and browser (where OpenTelemetry lacks details), we should be more liberal in setting standards.

---

For #392.

### CODE REVIEW

File: src/docs/sdk/performance/span-operations.mdx
Comment (maciejwalkowiak) [reply src/docs/sdk/performance/span-operations.mdx:87]: Yes.

**Comment by maciejwalkowiak** [comment]
@marandaneto with GraphQL on the server side, the transaction name is likely already fixed, so all queries/mutations go to the same transaction. Operation name is the way to distinguish what actually gets executed, which matters even more when there is more than one query sent with single request. I believe we should keep them dynamic.

**Comment by marandaneto** [comment]
> @marandaneto with GraphQL on the server side, the transaction name is likely already fixed, so all queries/mutations go to the same transaction. Operation name is the way to distinguish what actually gets executed, which matters even more when there is more than one query sent with single request. I believe we should keep them dynamic.

the problem is that we have a chart with an operation breakdown, if its dynamic, its not gonna be readable, maybe the dynamic part should go under the description, but we can talk about that after agreeing on a specification here.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:21]: Yeah this is a function of this being WIP. Will rename `Op` -> `Category` to make my intentions here a little more clear. 

The `Usage` column was just a collection of operations I've seen being used throughout the SDKs

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:37]: Yes I think the natural next step here is to build out a set of semantic conventions for ourselves on how you would build different integrations and what type of info (error -> context, tags, capture calls, performance -> transactions, spans, tags) you collect.

I think for now we keep this here as a starting point.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:63]: I think I might use `render` instead then.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:39]: Added as part of Browser operations

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:97]: You think they should follow `http.server` possibly? We could also special case `serverless`.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:77]: grabbed from https://github.com/getsentry/sentry-symfony/blob/5129f962a9b921f08fa2100fc5593cd6d8f06566/src/EventListener/TracingConsoleListener.php#L62, but this seems to be more related to running stuff on the command line. I'll move into web server (since using rails, django cmd line is pretty common operations).

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:31]: For sure. We've been using plain categories (with no actions or identifiers) for now. We can change this, but I think that's something we can come back to.

image

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:11]: Yeah we can decide on how to make individual platforms more consistent after we come to a general decision on the spec here.

**Review by AbhiPrasad — COMMENTED** [review]
Thanks everyone for your feedback/comments. I'm going to work to polish this up and add more examples (including snippets from different SDKs) so we can do another round of reviews.

File: src/docs/sdk/performance/span-operations.mdx
Comment (philipphofmann) [inline src/docs/sdk/performance/span-operations.mdx:114]: In a call, we discussed what to put in the description of the file span. We thought that filename + bytes count `file.txt (203 KB)` should go in the description and the full file path should go into data of the span as `data["filepath"] = "/absolute/example/file.txt"`. Should the bytes count be KB, MB, GB, ... ?

Can we document this here?

File: src/docs/sdk/performance/span-operations.mdx
Comment (philipphofmann) [inline src/docs/sdk/performance/span-operations.mdx:23]: Should we specify what should go into the description of the spans as well? I think it would be awesome to specify them precisely, so these are consistent as well.

**Review by philipphofmann — COMMENTED** [review]
Good job, @AbhiPrasad. I like the direction where this is going.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:27]: Yes I think this is a great idea, but out of the scope of this PR. I want to focus on getting the operations right, we can then take a look at descriptions next.

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:123]: > Can we document this here?

We should - but I'm not going to include it in this PR.

Instead we should be trying to develop our own set of semantic conventions (especially for desktop/mobile/browser - areas where open telemetry doesn't have anything). I plan to take this on after this PR gets merged in.

Comment (philipphofmann) [reply src/docs/sdk/performance/span-operations.mdx:123]: Okay, I created an issue so this doesn't get lost https://github.com/getsentry/develop/issues/463

File: src/docs/sdk/performance/span-operations.mdx
Comment (philipphofmann) [reply src/docs/sdk/performance/span-operations.mdx:27]: I created an issue for that https://github.com/getsentry/develop/issues/463

**Comment by AbhiPrasad** [comment]
I've updated the PR description with more details, and this is now ready for review. Any feedback is highly appreciated!

File: src/docs/sdk/performance/span-operations.mdx
Comment (romtsn) [inline src/docs/sdk/performance/span-operations.mdx:95]: Still a bit unclear what's the diff between `db.sql.query` and `db.query`? Should we just get rid of `db.sql.query`?

Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:95]: According to the outlined spec `sql` is an identifier for the `db` category. I'm of the opinion that both are fine to exist - `db.query` is just the more general one. 

The key takeaway here is that since both use the same category, `db`, they will be classified the same way in the product. Everything after the `db` prefix just provides additional context, SDK developers should feel free to choose whatever they think fits the best (as long as the category stays consistent).

Open to discussing further if you feel that we should take a more opinionated approach.

File: src/docs/sdk/performance/span-operations.mdx
Comment (rhcarvalho) [reply src/docs/sdk/performance/span-operations.mdx:11]: > I think Redis was an unfortunate example, it doesn't use SQL.

The Redis example also doesn't follow the `CATEGORY.ACTION` format we're just starting to introduce in the prose, so I'd suggest starting with something simpler, and following up with a more complex case.

```suggestion
Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, an instrumented HTTP server creates transaction spans named `http.server`, where `http` is the category, and `server` is the category identifier. Another example is a mobile app rendering a screen with a span operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.
```

Hmm not happy with `http.server`... I'd rather have something that follows the `CATEGORY.ACTION` format there.

Comment (rhcarvalho) [reply src/docs/sdk/performance/span-operations.mdx:11]: 🤔 💭 

There's both "category identifier" and "action identifier" on top of "category" and "action", but in the current shape of the spec we only see 1, 2 or 3 dot-separated items making up the operation name.

I wonder if there's a simpler way to refer to the parts. `CATEGORY[.CAT_IDENT].ACTION[.ACTION_IDENT]` is ambiguous to parse (for computers and humans), and so the 4 names for the parts feel like unnecessary complexity that we can't really make good use of.

For example, we cannot split up the `span.op` and render something descriptive in Sentry like "Operation: DB; Action: Insert".

File: src/docs/sdk/performance/span-operations.mdx
Comment (rhcarvalho) [inline src/docs/sdk/performance/span-operations.mdx:13]:
```suggestion
It's important to keep categories consistent between SDKs and integrations as they are used by Sentry in the [Operations Breakdown](https://docs.sentry.io/product/sentry-basics/tracing/event-detail/#operations-breakdown) feature. For example, both `db.init` and `db.query` will be categorized as database operations (`db`). The default operations breakdown config can be seen [here](https://github.com/getsentry/sentry/blob/809b7fe54c6f06cc1e4c503cf83ded896472a011/src/sentry/projectoptions/defaults.py#L74).
```

File: src/docs/sdk/performance/span-operations.mdx
Comment (rhcarvalho) [inline src/docs/sdk/performance/span-operations.mdx:15]: Needs verification

- [ ] We start out saying "Span operations are low cardinality attributes", and this sentence here relaxes that expectation, essentially accepting high-cardinality span operations, which sounds contradictory at a minimum, and might have negative performance consequences.

**Review by rhcarvalho — COMMENTED** [review]
A pass on the spec, ignored the examples for now.

File: src/docs/sdk/event-payloads/properties/op.mdx
Comment (rhcarvalho) [reply src/docs/sdk/event-payloads/properties/op.mdx:9]: `http.server` is probably a case of `CATEGORY.CATEGORY_IDENTIFIER`, a `CATEGORY.ACTION` example would fit better.

**Comment by rhcarvalho** [comment]
- [ ] Add a link from [/sdk/performance/](https://develop-git-abhi-span-operations.sentry.dev/sdk/performance/) to the [new Span Operations subpage](https://develop-git-abhi-span-operations.sentry.dev/sdk/performance/span-operations/)

File: src/docs/sdk/performance/span-operations.mdx
Comment (romtsn) [reply src/docs/sdk/performance/span-operations.mdx:95]: Ah alright, I thought it will affect the operation breakdown rendering in the UI, but if it's just there for context, I'm good with it, thanks

File: src/docs/sdk/performance/span-operations.mdx
Comment (k-fish) [reply src/docs/sdk/performance/span-operations.mdx:15]: The important point here isn't that it's generated, but that the span categories stay low cardinality. I think just recommending that if a span category **is** generated, that it should only use low cardinality inputs (a static array like how we create the ops breakdown measurements) and not something like an `id`, `url`, etc. 

Since `generated` is a fairly conflated term, I'd perhaps rewrite this to state that:
- It should be low cardinality
- If you **are** generating this somehow, take care to not use high cardinality inputs

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:2]: Opened https://github.com/getsentry/develop/issues/464 to track that

File: src/docs/sdk/performance/span-operations.mdx
Comment (rhcarvalho) [inline src/docs/sdk/performance/span-operations.mdx:5]:
```suggestion
Span operations are a short code identifying the type of operation the span is measuring. Span operations are low cardinality attributes - they should be as general as possible while still being human readable and useful. They should avoid including high cardinality data like IDs and URLs.
```

nit: We use `ID`/`IDs` in prose here and in docs.sentry.io.

**Comment by AbhiPrasad** [comment]
In the spirit of going step by step, I slightly changed the scope of the PR.

This PR now contains just a brief overview of operations and the table of the values we want to be used. This hopefully will help unblock the SDKs who just want the list of operations + encourage others to start adding to/refining this list.

I removed the note about operations structure (the stuff about `CATEGORY` and `ACTION` ) in https://github.com/getsentry/develop/pull/452/commits/c3c8428e05c2930e3a9b1f2711bfa0ec9620c0ce as that still needs to be discussed further. We can move those convos to another PR.

This is ready for another review

**Review by rhcarvalho — APPROVED** [review]
This is a good pass at documenting the current state.

**Comment by rhcarvalho** [comment]
> In the spirit of going step by step, I slightly changed the scope of the PR.

👍 

Reminder to check if you want `Fixes #392` in the PR description or not, as the issue itself is twofold.

File: src/docs/sdk/performance/span-operations.mdx
Comment (rhcarvalho) [inline src/docs/sdk/performance/span-operations.mdx:31]: Hmm I feel like I could live without this commit; but you adding it to this PR suggests the distinction between `Category` and `Usage` columns is not really necessary.

All we have as actual operations in SDKs are strings, and since we've dropped the intention to document how categories work we could as well have a single column:

| Operation | Description |
| --- | --- |
| ... | ... |

**Review by k-fish — APPROVED** [review]
This looks good to me, glad we added `ui.*` as I think it's something we need to be consistent in adding coverage for 👍

File: src/docs/sdk/performance/span-operations.mdx
Comment (AbhiPrasad) [reply src/docs/sdk/performance/span-operations.mdx:31]: Will address this in the next PR.

**Pull request merged**
